### PR TITLE
[Merged by Bors] - ConjTranspose multiplication of a matrix by itself is Positive Semidefinite

### PR DESCRIPTION
**Pull request by MohanadAhmed**
This PR provides two lemmas `isPosSemidef_conjTranspose_mul_self` and `isPosSemidef_self_mul_conjTranspose` stating that for any matrix $A$ the products $A^HA$ and $AA^H$ in the `IsROrC` fields are positive semidefinite.

Co-authored-by: Mohanad Ahmed <m.a.m.elhassan@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Matrix/PosDef.lean [hunk 1]
```diff
@@ -165,4 +166,25 @@ def InnerProductSpace.ofMatrix {M : Matrix n n 𝕜} (hM : M.PosDef) :
   InnerProductSpace.ofCore _
 #align matrix.inner_product_space.of_matrix Matrix.InnerProductSpace.ofMatrix
 
+section ConjTransposeMul
+
+variable {m : Type _} [Fintype m]
+open BigOperators
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/PosDef.lean:172]:
```suggestion
open scoped BigOperators
```

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/PosDef.lean:172]: Though you don't even use the notation, so I think you can drop this.

Comment (MohanadAhmed) [reply Mathlib/LinearAlgebra/Matrix/PosDef.lean:172]: Sure it is just for the goal display in infoview

File: Mathlib/LinearAlgebra/Matrix/PosDef.lean [hunk 2]
```diff
@@ -165,4 +166,25 @@ def InnerProductSpace.ofMatrix {M : Matrix n n 𝕜} (hM : M.PosDef) :
...
+  rw [← mulVec_mulVec, dotProduct_mulVec, vecMul_conjTranspose, star_star, dotProduct,
+    IsROrC.re_sum]
+  simpa [Pi.star_apply, IsROrC.star_def, IsROrC.mul_re, IsROrC.conj_re, IsROrC.conj_im, neg_mul,
+    sub_neg_eq_add] using
+  Finset.sum_nonneg (fun i => (fun _ => add_nonneg (mul_self_nonneg _) (mul_self_nonneg _)))
+
+/-- A matrix multiplied by its conjugate transpose is positive semidefinite -/
+theorem isPosSemidef_self_mul_conjTranspose (A : Matrix m n 𝕜) : Matrix.PosSemidef (A ⬝ Aᴴ) := by
+  nth_rewrite 1 [← conjTranspose_conjTranspose A]
+  apply isPosSemidef_conjTranspose_mul_self
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/PosDef.lean:186]: Does this work?
```suggestion
theorem isPosSemidef_self_mul_conjTranspose (A : Matrix m n 𝕜) : Matrix.PosSemidef (A ⬝ Aᴴ) :=
  isPosSemidef_conjTranspose_mul_self Aᴴ
```

Comment (MohanadAhmed) [reply Mathlib/LinearAlgebra/Matrix/PosDef.lean:186]: No

File: Mathlib/Data/IsROrC/Basic.lean [hunk 3]
```diff
@@ -291,6 +291,14 @@ theorem norm_ofReal (r : ℝ) : ‖(r : K)‖ = |r| :=
   norm_algebraMap' K r
 #align is_R_or_C.norm_of_real IsROrC.norm_ofReal
 
+@[simp, isROrC_simps]
+theorem re_sum (f : n → K) : IsROrC.re (∑ i in s, f i) = ∑ i in s, IsROrC.re (f i) := by
+  apply map_sum _ _
+
+@[simp, isROrC_simps]
+theorem im_sum (f : n → K) : IsROrC.im (∑ i in s, f i) = ∑ i in s, IsROrC.im (f i) := by
+  apply map_sum _ _
```
[the remark this answers is not in the document]
Comment (MohanadAhmed) [reply Mathlib/Data/IsROrC/Basic.lean:300]: I didn't use the simp aspect of it anyway as you can see it is in a rewrite

File: Mathlib/LinearAlgebra/Matrix/PosDef.lean (hunk 2 above)
Comment (MohanadAhmed) [reply Mathlib/LinearAlgebra/Matrix/PosDef.lean]: I changed it to `simpa [conjTranspose_conjTranspose] using isPosSemidef_conjTranspose_mul_self Aᴴ`

File: Mathlib/Data/IsROrC/Basic.lean (hunk 3 above)
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/Data/IsROrC/Basic.lean:300]: The difference is that `Complex.re` is a function but `IsROrC.re` is an additive map. I would guess your proof goes through fine if you replace `re_sum` with `map_sum`?

File: Mathlib/LinearAlgebra/Matrix/PosDef.lean [hunk 4]
```diff
@@ -165,4 +166,23 @@ def InnerProductSpace.ofMatrix {M : Matrix n n 𝕜} (hM : M.PosDef) :
...
+  refine ⟨ isHermitian_transpose_mul_self _, fun x => ?_ ⟩
+  rw [← mulVec_mulVec, dotProduct_mulVec, vecMul_conjTranspose, star_star, dotProduct,
+    IsROrC.re_sum]
+  simpa [Pi.star_apply, IsROrC.star_def, IsROrC.mul_re, IsROrC.conj_re, IsROrC.conj_im, neg_mul,
+    sub_neg_eq_add] using
+  Finset.sum_nonneg (fun i => (fun _ => add_nonneg (mul_self_nonneg _) (mul_self_nonneg _)))
+
+/-- A matrix multiplied by its conjugate transpose is positive semidefinite -/
+theorem isPosSemidef_self_mul_conjTranspose (A : Matrix m n 𝕜) : Matrix.PosSemidef (A ⬝ Aᴴ) :=
+  by simpa [conjTranspose_conjTranspose] using isPosSemidef_conjTranspose_mul_self Aᴴ
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/PosDef.lean:184]:
```suggestion
  by simpa only [conjTranspose_conjTranspose] using isPosSemidef_conjTranspose_mul_self Aᴴ
```

File: Mathlib/Data/IsROrC/Basic.lean (hunk 3 above)
[the remark this answers is not in the document]
Comment (MohanadAhmed) [reply Mathlib/Data/IsROrC/Basic.lean:300]: Your suggestion works, i.e. replace IsROrC.re_sum with map_sum. Do you want me to remove them?

File: Mathlib/LinearAlgebra/Matrix/PosDef.lean [hunk 5]
```diff
@@ -165,4 +166,22 @@ def InnerProductSpace.ofMatrix {M : Matrix n n 𝕜} (hM : M.PosDef) :
   InnerProductSpace.ofCore _
 #align matrix.inner_product_space.of_matrix Matrix.InnerProductSpace.ofMatrix
 
+section ConjTransposeMul
+
+variable {m : Type _} [Fintype m]
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/PosDef.lean:171]: Isn't there already an `m` at the top of the file?

Comment (MohanadAhmed) [reply Mathlib/LinearAlgebra/Matrix/PosDef.lean:171]: I think it is in a separate namespace. There are three namespaces in this file

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks!

**Comment by MohanadAhmed** [comment]
> ✌️ MohanadAhmed can now approve this pull request. To approve and merge a pull request, simply reply with `bors r+`. More detailed instructions are available [here](https://bors.tech/documentation/getting-started/#reviewing-pull-requests).

bors r+

**Pull request closed without merging**